patches/2025-03-10-create-primary-name-request-demand-factor.lua

```diff
@@ -0,0 +1,239 @@
+	end
+	local validFundFrom = utils.createLookupTable({ "any", "balance", "stakes" })
+	assert(validFundFrom[fundFrom], "Invalid fund from type. Must be one of: any, balance, stakes")
+end
+
+local function addPrimaryNameCounts(ioEvent)
+	ioEvent:addField("Total-Primary-Names", utils.lengthOfTable(primaryNames.getUnsafePrimaryNames()))
+	ioEvent:addField("Total-Primary-Name-Requests", utils.lengthOfTable(primaryNames.getUnsafePrimaryNameRequests()))
+end
+
+local function adjustSuppliesForFundingPlan(fundingPlan, rewardForInitiator)
+	if not fundingPlan then
+		return
+	end
+	rewardForInitiator = rewardForInitiator or 0
+	local totalActiveStakesUsed = utils.reduce(fundingPlan.stakes, function(acc, _, stakeSpendingPlan)
+		return acc + stakeSpendingPlan.delegatedStake
+	end, 0)
+	local totalWithdrawStakesUsed = utils.reduce(fundingPlan.stakes, function(acc, _, stakeSpendingPlan)
+		return acc
+			+ utils.reduce(stakeSpendingPlan.vaults, function(acc2, _, vaultBalance)
+				return acc2 + vaultBalance
+			end, 0)
+	end, 0)
+	LastKnownStakedSupply = LastKnownStakedSupply - totalActiveStakesUsed
+	LastKnownWithdrawSupply = LastKnownWithdrawSupply - totalWithdrawStakesUsed
+	LastKnownCirculatingSupply = LastKnownCirculatingSupply - fundingPlan.balance + rewardForInitiator
+end
+
+local function addResultFundingPlanFields(ioEvent, result)
+	ioEvent:addFieldsWithPrefixIfExist(result.fundingPlan, "FP-", { "balance" })
+	local fundingPlanVaultsCount = 0
+	local fundingPlanStakesAmount = utils.reduce(
+		result.fundingPlan and result.fundingPlan.stakes or {},
+		function(acc, _, delegation)
+			return acc
+				+ delegation.delegatedStake
+				+ utils.reduce(delegation.vaults, function(acc2, _, vaultAmount)
+					fundingPlanVaultsCount = fundingPlanVaultsCount + 1
+					return acc2 + vaultAmount
+				end, 0)
+		end,
+		0
+	)
+	if fundingPlanStakesAmount > 0 then
+		ioEvent:addField("FP-Stakes-Amount", fundingPlanStakesAmount)
+	end
+	if fundingPlanVaultsCount > 0 then
+		ioEvent:addField("FP-Vaults-Count", fundingPlanVaultsCount)
+	end
+	local newWithdrawVaultsTallies = utils.reduce(
+		result.fundingResult and result.fundingResult.newWithdrawVaults or {},
+		function(acc, _, newWithdrawVault)
+			acc.totalBalance = acc.totalBalance
+				+ utils.reduce(newWithdrawVault, function(acc2, _, vault)
+					acc.count = acc.count + 1
+					return acc2 + vault.balance
+				end, 0)
+			return acc
+		end,
+		{ count = 0, totalBalance = 0 }
+	)
+	if newWithdrawVaultsTallies.count > 0 then
+		ioEvent:addField("New-Withdraw-Vaults-Count", newWithdrawVaultsTallies.count)
+		ioEvent:addField("New-Withdraw-Vaults-Total-Balance", newWithdrawVaultsTallies.totalBalance)
+	end
+	adjustSuppliesForFundingPlan(result.fundingPlan, result.returnedName and result.returnedName.rewardForInitiator)
+end
+
+--- @param ioEvent ARIOEvent
+--- @param primaryNameResult CreatePrimaryNameResult|PrimaryNameRequestApproval
+local function addPrimaryNameRequestData(ioEvent, primaryNameResult)
```

Comment:
unfortunately, these all need to be added as they are not available in the global scope